File: prep/editor.py
# # HalHello World
# # again and again

class Editor:

    def __init__(self) -> None:
        self.buf = ""
        self.pos = 0 # next insert position
    
    def insert(self, c):
        self.buf = self.buf[:self.pos] + c + self.buf[self.pos:]
        self.pos += 1

    def goto_line_start(self):
        while self.pos > 0:
            if self.buf[self.pos-1] == '\n':
                break
            self.pos -= 1

    def goto_line_end(self):
        while self.pos < len(self.buf)-1:
            if self.buf[self.pos] == '\n':
                break
            self.pos += 1
    
    def left(self):
        if self.pos > 0:
            self.pos -= 1
    
    def right(self):
        if self.pos < len(self.buf)-1:
            self.pos += 1

    def count_to_start(self):
        p = self.pos
        c = 0
        while p > 0:
            if self.buf[p-1] == '\n':
                break
            p -= 1
            c += 1
        return c

    def down(self):
        offs = self.count_to_start()        
        self.goto_line_end()
        self.right()
        for _ in range(offs):
            self.right()


    def up(self):
        offs = self.count_to_start()        
        self.goto_line_start()
        self.left()
        self.goto_line_start()
        for _ in range(offs):
            self.right()

    def show(self):
        print(self.buf)

ed = Editor()
for c in "Hello World":
    ed.insert(c)
ed.insert("\n")
for c in "you suckers":
    ed.insert(c)

ed.goto_line_start()
print(ed.pos)
ed.left()
print(ed.pos)
ed.goto_line_start()
print(ed.pos)
ed.right()
ed.right()
print(ed.pos)
ed.insert("-")
ed.left()
print('pos:',ed.pos)
ed.down()
ed.insert("-")
ed.right()
ed.insert("*")
ed.left()
ed.up()
ed.insert("*")
print(ed.pos)
# ed.insert("\n")
# for c in "again and again":
#     ed.insert(c)
# ed.up()
# ed.show()
# ed.insert("!")
# ed.show()
# ed.down()
# ed.insert("?")
ed.show()

File: prep/editor.cpp
#include <bits/stdc++.h>
#include <cstddef>
#include <cstring>
using namespace std;

struct Editor {

    string buf = "";
    int pos = 0;

    void insert(char c) {
        buf.insert(pos, 1, c);
        pos += 1;
    }

    void show() {
        cout << buf;
    }

    // move "after" the \n that separated the lines
    void goto_line_start() {
        while (pos > 0) {
            if (buf[pos-1] == '\n') {
                break;
            }
            pos--;
        }
    }

    void goto_line_end() {
        while (pos < buf.size()-1) {
            if (buf[pos] == '\n') {
                break;
            }
            pos++;
        }
    }

    int count_to_line_start() {
        int p = pos;
        int c = 0;
        while (p > 0) {
            if (buf[p-1] == '\n') {
                break;
            }
            p--;
            c++;
        }
        return c;
    }

    void left() {
        if (pos > 0) pos--;
    }

    void right() {
        if (pos < buf.size()-1) pos++;
    }

    void down() {
        int offset = count_to_line_start();
        goto_line_end();
        right();
        for (int i=0; i<offset; i++) {
            right();
        }
    }

    void up() {
        int offset = count_to_line_start();
        goto_line_start();
        left();
        goto_line_start();
        for (int i=0; i<offset; i++) {
            right();
        }
    }
};


int main() {
    Editor ed;
    for (char c : "Hello World") {
        ed.insert(c);
    }
    ed.insert('\n');
    for (char c : "you suckers") {
        ed.insert(c);
    }

    ed.goto_line_start();
    cout << ed.pos << endl;
    cout << ed.buf[ed.pos] << endl;
    ed.left();
    cout << ed.pos << endl;
    ed.goto_line_start();
    cout << ed.pos << endl;
    ed.right();
    ed.right();
    cout << ed.pos << endl;
    ed.insert('-');
    ed.left();
    // print('pos:',ed.pos)
    ed.down();
    ed.insert('-');
    ed.right();
    ed.insert('*');
    ed.left();
    ed.up();
    ed.insert('*');
    ed.show();
    return 0;
}

